### PR TITLE
Add cars filtering

### PR DESCRIPTION
**Pull request by R0ixy**
Asana - https://app.asana.com/0/1202525679663009/1202720801120415/f

### CODE REVIEW

File: backend/src/helpers/cars/api_autoria.helper.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { ENV } from '@common/enums/app/env.enum';
+import { AutoriaRequestParams } from '@common/types/types';
+import axios from 'axios';
+
+const URL = 'https://developers.ria.com/auto/search';
```
Comment (nikita-remeslov) [inline backend/src/helpers/cars/api_autoria.helper.ts]: Maybe better move it to env? @oleksandr-danylchenko

File: backend/src/common/types/cars/autoria_request_params.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+interface AutoriaRequestParams {
+  category_id: number;
+  price_ot?: string;
+  price_do?: string;
+  powerFrom?: string;
+  powerTo?: string;
+  power_name?: number;
+  engineVolumeFrom?: string;
+  engineVolumeTo?: string;
```
Comment (nikita-remeslov) [inline backend/src/common/types/cars/autoria_request_params.ts]: Are you sure you need that many optional keys?
In your case for request, we can use `Partial<AutoriaRequestParams>`

File: backend/src/helpers/cars/api_autoria.helper.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+const getCarsAutoRia = async (
+  requestParams: AutoriaRequestParams,
```
Comment (nikita-remeslov) [inline backend/src/helpers/cars/api_autoria.helper.ts]:
```suggestion
  requestParams: Partial<AutoriaRequestParams>,
```

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/helpers/cars/api_autoria.helper.ts (hunk 1 above)
Comment (oleksandr-danylchenko) [reply backend/src/helpers/cars/api_autoria.helper.ts]: I think it's okay to leave it in the source code. It won't change depending on ENV and I have high hopes that the autoria api URL won't change in a next month

File: backend/src/common/types/cars/autoria_request_params.ts (hunk 2 above)
Comment (oleksandr-danylchenko) [reply backend/src/common/types/cars/autoria_request_params.ts]: :arrow_double_up: :+1:

File: backend/src/controllers/cars/cars.controller.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+const carsSearch = async (
+  req: TypedRequestQuery<CarsSearchParams>,
+  res: Response,
+  next: NextFunction,
+): Promise<void> => {
+  try {
+    const carsData = await carsService.carsSearch(req.query);
+    const AutoRiaCarsData = await getCarsAutoRia(carsData);
+
+    res.json(AutoRiaCarsData);
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/cars/cars.controller.ts]: Please use `camelCase`

File: backend/src/controllers/cars/cars.controller.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  } catch (error) {
+    if ((error as Error)?.message === 'NotFoundError') {
+      res.json([]);
+    } else {
+      next(error);
+    }
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/cars/cars.controller.ts]: You should use `if (error instanceof Error)` check

File: backend/src/services/cars/carsSearch.service.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
+import { CarsSearchParams } from '@autoline/shared';
```
Comment (oleksandr-danylchenko) [inline backend/src/services/cars/carsSearch.service.ts]: All the files should be named in `kebab-case`. Please rename the files named in `snake_case` and `camelCase`

File: backend/src/services/cars/carsSearch.service.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    model_from_db.forEach((model, index) => {
+      Object.assign(result, {
+        [`body_type[${index}]`]: model.body_type?.autoria_code,
+        [`marka_id[${index}]`]: model.brand?.autoria_code,
+        [`model_id[${index}]`]: data.modelId ? model.autoria_code : undefined,
+        [`s_yers[${index}]`]: data.yearStart,
+        [`po_yers[${index}]`]: data.yearEnd,
+      });
+    });
```
Comment (oleksandr-danylchenko) [inline backend/src/services/cars/carsSearch.service.ts]: Could you explain what it does?

File: backend/src/services/cars/carsSearch.service.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+  await getData(data.colorId, 'color_id', prisma.color, result);
+  await getData(
+    data.transmissionTypeId,
+    'gear_id',
+    prisma.transmission_Type,
+    result,
+  );
+  await getData(data.fuelTypeId, 'type_id', prisma.fuel_Type, result);
+  await getData(data.drivetrainId, 'drive_id', prisma.drivetrain, result);
```
Comment (oleksandr-danylchenko) [inline backend/src/services/cars/carsSearch.service.ts]: Why the method which is named `getSmth` doesn't return the value? Mutating the parameter is a cursed anti-pattern straight from 1998

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]
Some simplification and comments would be appreciated

File: backend/src/services/cars/carsSearch.service.ts (hunk 7 above)
Comment (R0ixy) [reply backend/src/services/cars/carsSearch.service.ts]: It's just the way autoria asks to specify params. If you want to search multiply models or brands you need to add them this way `marka_id[0]=423&marka_id[1]=222` and so on. And if you want to search by years you need to duplicate it for every model/brand. That's actually what this code does

**Review by oleksandr-danylchenko — APPROVED** [review]

**Pull request merged**